common/record_logger.hpp

```diff
@@ -0,0 +1,317 @@
+#pragma once
+
+#include <any>
+#include <atomic>
+#include <unordered_map>
+#include <sstream>
+#include <vector>
+#include <memory>
+#include "phonebook.hpp"
+
+namespace ILLIXR {
+
+	/**
+	 * @brief Schema of each record.
+	 *
+	 * name_ should be globally unique.
+	 */
+	class record_header {
+	public:
+		record_header(std::string name_, std::vector<std::pair<std::string, const std::type_info&>> columns_)
+			: id{std::hash<std::string>{}(name_)}
+			, name{name_}
+			, columns{columns_}
+		{ }
+
+		/**
+		 * @brief Compares two schemata.
+		 */
+		bool operator==(const record_header& other) const {
+			// Check pointer first
+			if (this == &other) {
+				return true;
+			}
+
+			if (false
+				|| name != other.name
+				|| columns.size() != other.columns.size()
+				|| id != other.id
+				) {
+				return false;
+			}
+			for (std::size_t i = 0; i < columns.size(); ++i) {
+				if (columns[i] != other.columns[i]) {
+					return false;
+				}
+			}
+			return true;
+		}
+		bool operator!=(const record_header& other) const {
+			return !(*this == other);
+		}
+		std::size_t get_id() const { return id; }
+		const std::string& get_name() const { return name; }
+		const std::string& get_column_name(unsigned column) const { return columns[column].first; }
+		const std::type_info& get_column_type(unsigned column) const { return columns[column].second; }
+		unsigned get_columns() const { return columns.size(); }
+		std::string to_string() const {
+			std::string ret = std::string{"record_header "} + name + std::string{" { "};
+			for (const auto& pair : columns) {
+				ret += std::string{pair.second.name()} + std::string{" "} + pair.first + std::string{"; "};
+			}
+			ret.erase(ret.size() - 2);
+			ret += std::string{" }"};
+			return ret;
+		}
+
+	private:
+		std::size_t id;
+		std::string name;
+		const std::vector<std::pair<std::string, const std::type_info&>> columns;
+	};
+
+	/**
+	 * @brief A helper class that lets one dynamically determine if some data gets used.
+	 *
+	 * When a use_taint gets copied, the original is considered used and the new one is considered unused.
+	 */
+	class use_taint {
```

Comment:
Nit: Can we use another word instead of taint? `use_taint` doesn't obviously describe what the class/function does.

Reply:
I'm alluding to [this](https://en.wikipedia.org/wiki/Taint_checking); I will look for a better word.